lib/api/CBoostedTreeInferenceModelBuilder.cc

```diff
@@ -0,0 +1,193 @@
+    ml::maths::CBoostedTreeNode::TOptionalNodeIndex rightChild) {
+    auto ensemble{static_cast<CEnsemble*>(m_Definition.trainedModel().get())};
+    // use dynamic cast to prevent using wrong type of trained models
+    auto tree = dynamic_cast<CTree*>(ensemble->trainedModels().back().get());
+    if (tree == nullptr) {
+        HANDLE_FATAL(<< "Internal error. Tree points to a nullptr.")
+    }
+    tree->treeStructure().emplace_back(tree->size(), splitValue, assignMissingToLeft, nodeValue,
+                                       splitFeature, leftChild, rightChild, gain);
+}
+
+CBoostedTreeInferenceModelBuilder::CBoostedTreeInferenceModelBuilder(
+    TStringVec fieldNames,
+    std::size_t dependentVariableColumnIndex,
+    const TStringSizeUMapVec& categoryNameMap) {
+    // filter filed names containing only "."
+    fieldNames.erase(std::remove(fieldNames.begin(), fieldNames.end(), "."),
+                     fieldNames.end());
+    // filter dependent variable field name
+    assert(dependentVariableColumnIndex < fieldNames.size());
+    fieldNames.erase(fieldNames.begin() +
+                     static_cast<std::ptrdiff_t>(dependentVariableColumnIndex));
+    m_FieldNames = fieldNames;
+
+    this->categoryNameMap(categoryNameMap);
+    m_Definition.fieldNames(fieldNames);
+    m_Definition.trainedModel(std::make_unique<CEnsemble>());
+    m_Definition.typeString(INFERENCE_MODEL);
+}
+
+CBoostedTreeInferenceModelBuilder::TStringDoubleUMap
+CBoostedTreeInferenceModelBuilder::encodingMap(std::size_t inputColumnIndex,
+                                               const TDoubleVec& map_) {
+    TStringDoubleUMap map;
+    for (std::size_t categoryUInt = 0; categoryUInt < map_.size(); ++categoryUInt) {
+        std::string category{m_ReverseCategoryNameMap[inputColumnIndex][categoryUInt]};
+        map.emplace(category, map_[categoryUInt]);
+    }
+    return map;
+}
+
+void CBoostedTreeInferenceModelBuilder::categoryNameMap(
+    const CInferenceModelDefinition::TStringSizeUMapVec& categoryNameMap) {
+    m_ReverseCategoryNameMap.reserve(categoryNameMap.size());
+    for (const auto& categoryNameMapping : categoryNameMap) {
+        if (categoryNameMapping.empty() == false) {
+            TSizeStringUMap map;
+            for (const auto& categoryMappingPair : categoryNameMapping) {
+                map.emplace(categoryMappingPair.second, categoryMappingPair.first);
+            }
+            m_ReverseCategoryNameMap.emplace_back(std::move(map));
+        } else {
+            m_ReverseCategoryNameMap.emplace_back();
+        }
+    }
+}
+
+CInferenceModelDefinition& CBoostedTreeInferenceModelBuilder::definition() {
+    return m_Definition;
+}
+
+void CRegressionInferenceModelBuilder::setTargetType() {
+    this->definition().trainedModel()->targetType(CTrainedModel::ETargetType::E_Regression);
+}
+
+CRegressionInferenceModelBuilder::CRegressionInferenceModelBuilder(
+    const CBoostedTreeInferenceModelBuilder::TStringVec& fieldNames,
```

Comment:
Since the type is visible via inheritance
```suggestion
    const TStringVec& fieldNames,
```
and three other cases below

Reply:
Get fixed after merge.